Move the DOM-based search into the javac bundle

robstryker opened the pull request.

mickaelistria committed:
Make MatchLocator capable of working with DOM

Extra strategy in MatchLocator to use DOM, and tweaks in various
Locators to properly handle DOM. Tests in JavaSearchTests help a lot
(currently ~130/270 passing)

It basically clones and adapt some methods and adapt them to DOM,
A new flag is introduced to decide the strategy.

Done-ish (often with remaining issues)
* FieldLocator
* LocalVariableLocator
* MethodLocator
* SuperTypeReferenceLocator
* TypeDeclarationLocator
* TypeParameterLocator
* TypeReferenceLocator
TODO:
* AndLocator
* ModuleLocator
* PackageDeclarationLocator
* PackageReferenceLocator
* OrLocator
* VariableLocator

Make some match locator work with DOM

Fix NPE, convert 4 errors to fails

Signed-off-by: Rob Stryker <stryker@redhat.com>

Handle match files in the order they arrived

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix testBug251827b and 5 others - import declarations java element must be found

Signed-off-by: Rob Stryker <stryker@redhat.com>

Move the DOM-based search into the javac bundle

Signed-off-by: Rob Stryker <rob@oxbeef.net>

Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Move the DOM-based search into the javac bundle

Signed-off-by: Rob Stryker <stryker@redhat.com>

robstryker committed:
Allow overriding the Java search logic with a custom implementation

Signed-off-by: Rob Stryker <rob@oxbeef.net>

mickaelistria commented:
That can make a lot of sense for now.
Was the suggested extension point already submitted upstream?

robstryker commented:
Extension point is already submitted upstream but not yet merged. 

This branch will likely need a careful application when being merged, so please (when appropriate) just give your approval but do not merge it yourself. I will handle it. 

This PR is slightly different from the upstream one, and we may need to reorder the commits so as to minimize the affects.

robstryker commented:
This was a bit messy, but I ended up removing the following commits from history: 

* 0595211439 - Fix testBug251827b and 5 others - import declarations java element must be found (25 minutes ago) <Rob Stryker>
* fa7d4925bf - Handle match files in the order they arrived
* 65a7ea1791 - Fix NPE, convert 4 errors to fails
* 8ce1472b8b - Make some match locator work with DOM (25 minutes ago) <Rob Stryker>
* 028f1481ea - Make MatchLocator capable of working with DOM (10 hours ago) <Mickael Istria>

Then I cherry-picked them as part of this PR, and then I squashed some together. 

So this PR included the commit still up for submission at jdt.core main repo (extension point) and the next 6 commits rebased, cleaned up, and squashed. 

So... that... looks good now.
